Adds support for loop unrolling within range loops

Rohit-Kakodkar opened the pull request:
## Description

Related to https://github.com/kokkos/kokkos/issues/8125.
- Adds a `LoopUnroll` trait within `Experimental` namespace 
- CUDA implementation

## Benchmark

Minimal benchmark for sanity check
```cpp
#include <Kokkos_Core.hpp>

int main(int argc, char *argv[]) {
  Kokkos::initialize(argc, argv);
  {
    int N = argc > 1 ? atoi(argv[1]) : 10000000;
    int V = argc > 2 ? atoi(argv[2]) : 0;
    Kokkos::View<int *> a("A", N);

    const int warmup = 10;

    auto reset_lambda = KOKKOS_LAMBDA(int i) { a(i) = 0; };

    auto lambda = KOKKOS_LAMBDA(int i) { a(i) = V; };

    int N4 = N / 4;
    auto lambda4 = KOKKOS_LAMBDA(int i) {
      a(i) = V;
      a(i + N4) = V;
      a(i + 2 * N4) = V;
      a(i + 3 * N4) = V;
    };

    auto lambda4_unrolled = KOKKOS_LAMBDA(int i) {
#pragma unroll
      for (std::size_t j = 0; j < 4; ++j) {
        a(i + j * N4) = V;
      }
    };

    for (int run = 0; run < warmup; ++run) {
      Kokkos::deep_copy(a, V);
      Kokkos::parallel_for(N, lambda);

      Kokkos::parallel_for(N4, lambda4);
      Kokkos::parallel_for(N4, lambda4_unrolled);
      Kokkos::parallel_for(
          Kokkos::RangePolicy<Kokkos::Experimental::LoopUnroll<4>>(0, N),
          lambda);
    }

    // Reset the array to zero before measuring parallel_for with unrolling
    Kokkos::parallel_for(N, reset_lambda);
    Kokkos::fence();

    Kokkos::Timer timer;
    Kokkos::deep_copy(a, V);
    double time_dc = timer.seconds();

    // Reset the array to zero before measuring parallel_for
    Kokkos::parallel_for(N, reset_lambda);
    Kokkos::fence();

    timer.reset();
    Kokkos::parallel_for(N, lambda);
    Kokkos::fence();
    double time_pf = timer.seconds();

    // Reset the array to zero before measuring parallel_for with unrolling
    Kokkos::parallel_for(N, reset_lambda);
    Kokkos::fence();

    timer.reset();
    Kokkos::parallel_for(N / 4, lambda4);
    Kokkos::fence();
    double time_pf_opt = timer.seconds();

    // Reset the array to zero before measuring parallel_for with unrolling
    Kokkos::parallel_for(N, reset_lambda);
    Kokkos::fence();

    timer.reset();
    Kokkos::parallel_for(N / 4, lambda4_unrolled);
    Kokkos::fence();
    double time_pf_unroll = timer.seconds();

    // Reset the array to zero before measuring parallel_for with unrolling
    Kokkos::parallel_for(N, reset_lambda);
    Kokkos::fence();

    timer.reset();
    Kokkos::parallel_for(
        Kokkos::RangePolicy<Kokkos::Experimental::LoopUnroll<4>>(0, N), lambda);
    Kokkos::fence();
    double time_pf_unroll2 = timer.seconds();

    auto h_a = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), a);

    // Check results
    for (int i = 0; i < N; ++i) {
      if (h_a(i) != V) {
        printf("Error at index %d: expected %d, got %d\n", i, V, a(i));
      }
    }

    printf("Deep Copy: %lf\nParallel For: %lf\nManual Unroll: %lf\nPragma "
           "unroll: %lf\nUnroll Trait: %lf\n",
           time_dc * 1.e6, time_pf * 1.e6, time_pf_opt * 1.e6,
           time_pf_unroll * 1.e6, time_pf_unroll2 * 1.e6);
  }
  Kokkos::finalize();
}
```

Results: 
```
Deep Copy: 44.514000
Parallel For: 118.914000
Manual Unroll: 43.442000
Pragma unroll: 42.389000
Unroll Trait: 43.362000
```

Rohit-Kakodkar committed:
Added loop unroll trait | CUDA implementation

Rombur commented on core/src/traits/Kokkos_LoopUnrollTrait.hpp:
```diff
@@ -0,0 +1,86 @@
+    static constexpr bool loop_unroll_is_defaulted = false;
+
+    static_assert(base_t::loop_unroll_is_defaulted,
+                  "Kokkos Error: More than one launch_bounds given");
```
You need to update the error message

Rombur commented:
You need to fix the format or our test suite won't run

dalg24 commented on core/src/traits/Kokkos_LoopUnrollTrait.hpp:
```diff
@@ -0,0 +1,86 @@
+
+}  // end namespace Kokkos
+
+#endif  // KOKKOS_KOKKOS_LAUNCHBOUNDSTRAIT_HPP
```
```suggestion
#endif
```

dalg24 replied:
or
```suggestion
#endif  // KOKKOS_KOKKOS_LOOPUNROLLTRAIT_HPP
```

dalg24 commented on core/src/impl/Kokkos_AnalyzePolicy.hpp:
```diff
@@ -31,6 +31,7 @@
 #include <traits/Kokkos_ScheduleTrait.hpp>
 #include <traits/Kokkos_WorkItemPropertyTrait.hpp>
 #include <traits/Kokkos_WorkTagTrait.hpp>
+#include <traits/Kokkos_LoopUnrollTrait.hpp>
```
All the header includes above are in alphabetic order.  Please move it after `Kokkos_LaunchBoundsTrait.hpp`.

dalg24 commented on core/src/traits/Kokkos_LoopUnrollTrait.hpp:
```diff
@@ -0,0 +1,86 @@
+
+namespace Kokkos {
+
+namespace Experimental {
```
Prefer a nested namespace definition
```suggestion
namespace Kokkos::Experimental {
```

dalg24 reviewed (commented):
Obviously we will need to add testing as well.
Did you look at performance?  We could also add a benchmark.

Rohit-Kakodkar commented:
> Obviously we will need to add testing as well. Did you look at performance? We could also add a benchmark.

Agreed! I ran a minimal benchmarking example, see above. Can add it to benchmarking / testing suite.

Rohit-Kakodkar commented:
> You need to fix the format or our test suite won't run

Do you use `pre-commit` (or something else) for formatting?

Rohit-Kakodkar committed:
Fixed formatting issues

Rohit-Kakodkar committed:
Applied clang format

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rohit-Kakodkar committed:
Negative loop ranges

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

dalg24 commented on core/src/traits/Kokkos_LoopUnrollTrait.hpp:
```diff
@@ -0,0 +1,82 @@
+  using loop_unroll = LoopUnroll;
+  using type        = LoopUnroll<LoopUnrollFactor>;
+  constexpr static unsigned int unroll_factor{
+      LoopUnrollFactor};  // Default unroll factor is 1
```
We probably want to static assert the the unrolling factor is not zero

Rohit-Kakodkar committed:
static assertion on loop factor

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

masterleinad commented:
```C++
#include <Kokkos_Core.hpp>

int main(int argc, char *argv[]) {
 Kokkos::initialize(argc, argv);
 {
   int N = argc > 1 ? atoi(argv[1]) : 10000000;
   int V = argc > 2 ? atoi(argv[2]) : 0;
   Kokkos::View<int *> a("A", N);

   const int warmup = 10;

   auto lambda = KOKKOS_LAMBDA(int i) { a(i) = V; };

   for (int c=1; c<=16; c*=2) {

   int N4 = N / c;

   auto lambda4_unrolled = KOKKOS_LAMBDA(int i) {
     for (std::size_t j = 0; j < c; ++j) {
       a(i + j * N4) = V;
     }
   };

   for (int run = 0; run < warmup; ++run) {
     Kokkos::deep_copy(a, V);
     Kokkos::parallel_for(N, lambda);
     Kokkos::parallel_for(N4, lambda4_unrolled);
   }

   Kokkos::fence();

   Kokkos::Timer timer;
   Kokkos::deep_copy(a, V);
   double time_dc = timer.seconds();

   timer.reset();
   Kokkos::parallel_for(N, lambda);
   Kokkos::fence();
   double time_pf = timer.seconds();

   timer.reset();
   Kokkos::parallel_for(N / c, lambda4_unrolled);
   Kokkos::fence();
   double time_pf_unroll = timer.seconds();
   printf("%d:\n Deep Copy: %lf\nParallel For: %lf\nPragma unroll: %lf\n\n", c, time_dc * 1.e6, time_pf * 1.e6,
          time_pf_unroll * 1.e6);
   }
 }
 Kokkos::finalize();
}
```
gives me essentially the same performance
```
1:
Deep Copy: 30.908000
Parallel For: 69.650000
Pragma unroll: 69.861000

2:
Deep Copy: 32.039000
Parallel For: 70.421000
Pragma unroll: 41.056000

4:
Deep Copy: 33.393000
Parallel For: 69.579000
Pragma unroll: 31.178000

8:
Deep Copy: 32.140000
Parallel For: 69.519000
Pragma unroll: 31.499000

16:
Deep Copy: 30.767000
Parallel For: 69.470000
Pragma unroll: 31.840000
```
hinting that using the "unroll" pragma might not be as important here as simply making sure that each thread gets multiple work items. This also matches results in https://github.com/kokkos/kokkos/issues/8033.
This is pretty much what chunk size is designed for (for host backends; for SYCL it currently forces the group size) and I would want to explore if we can give it a similar meaning for GPU backends (or if it really needs to be a compile-time constant). Whether we use `pragma unroll` then is somewhat of an orthogonal question.

Rombur commented:
@masterleinad  In your example the different values of `c` are known at compile time, so the compiler can do all the optimization it wants. `chunk_size` is a runtime value which is harder for the compiler to optimize. Can you change the code and have `c` has an input parameter.

Rohit-Kakodkar commented:
Running a small test. I modified the implementation to check if the value needs to be a compile-time constant. Below is `git diff` so that cuda implementation uses chunk size 

```cpp
diff --git a/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp b/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp
index 0bcb4a171..a44761483 100644
--- a/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp
+++ b/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp
@@ -71,7 +71,7 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::Cuda> {
   Policy const& get_policy() const { return m_policy; }
 
   inline __device__ void operator()() const {
-    constexpr unsigned int loop_unroll_factor = LoopUnroll::unroll_factor;
+    const auto loop_unroll_factor         = m_policy.chunk_size();
     const auto work_stride                    = Member(blockDim.y) * gridDim.x;
     const Member work_end                     = m_policy.end();
 
@@ -85,9 +85,9 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::Cuda> {
 // Unroll the loop
 #pragma unroll
       for (unsigned int i = 0;
-           i < work_stride * loop_unroll_factor && iwork + i < work_end;
+           i < work_stride * loop_unroll_factor && iwork + static_cast<Member>(i) < work_end;
            i += work_stride) {
-        this->template exec_range<WorkTag>(iwork + i);
+        this->template exec_range<WorkTag>(iwork + static_cast<Member>(i));
       }
     }
   }

--- a/core/src/Kokkos_ExecPolicy.hpp
+++ b/core/src/Kokkos_ExecPolicy.hpp
@@ -188,7 +188,7 @@ class RangePolicy : public Impl::PolicyTraits<Properties...> {
 
  public:
   /** \brief return chunk_size */
-  inline member_type chunk_size() const { return m_granularity; }
+  constexpr KOKKOS_INLINE_FUNCTION member_type chunk_size() const { return m_granularity; }
```

And updated the benchmark in the description. 

```cpp
@@ -80,7 +80,7 @@ int main(int argc, char *argv[]) {
 
     timer.reset();
     Kokkos::parallel_for(
-        Kokkos::RangePolicy<Kokkos::Experimental::LoopUnroll<4>>(0, N), lambda);
+        Kokkos::RangePolicy(0, N, Kokkos::ChunkSize(4)), lambda);
     Kokkos::fence();
     double time_pf_unroll2 = timer.seconds();
```

I get the following results on A100 
```
Deep Copy: 29.616000
Parallel For: 73.088000
Manual Unroll: 27.663000
Pragma unroll: 27.962000
Unroll Trait: 83.828000
```

It seems like it needs to be a compile time constant. I am assuming we could still use `ChunkSize` for this purpose as long as we make those functions `constexpr` valid. Agreed that `#pragma unroll` might be redundant

Rombur commented on core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp:
```diff
@@ -70,22 +71,31 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::Cuda> {
-      this->template exec_range<WorkTag>(iwork);
+// Unroll the loop
+#pragma unroll
+      for (unsigned int i = 0; i < work_stride * loop_unroll_factor &&
```
There is a bug in the current code if `workstride = 2^32`. In this case,  `i == i + work_stride` and the loop never ends. Use `Member` instead of `unsigned int`.

Rohit-Kakodkar replied:
Latest benchmark with above change: 
```
Deep Copy: 28.844000
Parallel For: 72.928000
Manual Unroll: 27.472000
Pragma unroll: 27.872000
Unroll Trait: 27.972000
```

Rohit-Kakodkar committed:
Updated loop factor to Member

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rohit-Kakodkar committed:
Updated loop unroll to index_type

Rohit-Kakodkar committed:
Removed unnacessary member cast

Rohit-Kakodkar committed:
Adds tests for loop unrolling

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rohit-Kakodkar committed:
applied clang format

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rohit-Kakodkar committed:
Minor fixes

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rohit-Kakodkar committed:
Fixed assert

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rohit-Kakodkar committed:
Fixes loop range (missing remainder)

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rohit-Kakodkar committed:
Fixed member casting

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rohit-Kakodkar commented:
Hi @Rombur, @dalg24, I added the suggested tests. Seems like some of tests were timed out. I am not sure why [this one](https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-8164/13/pipeline-overview/log?nodeId=403) failed - weird since I didn't change the `OPENMPTARGET` backend.

masterleinad reviewed (dismissed):
As discussed before, I'd like to see some documentation for this feature that clarifies what a user can expect across all backends and how it interacts with setting the chunk size.

Rohit-Kakodkar commented:
> As discussed before, I'd like to see some documentation for this feature that clarifies what a user can expect across all backends and how it interacts with setting the chunk size.

Currently, this PR only adds unrolling support for CUDA backends. I was planning on updating each backend separately. So, currently, there shouldn't be any interaction with the chunk size, correct?  Where should I update the documentation?

Rombur commented:
@Rohit-Kakodkar don't worry about OpenMP-Target. There is an issue when using CUDA clang
```
/var/jenkins/workspace/Kokkos_PR-8164/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp:86:9: error: unknown pragma ignored [clang-diagnostic-unknown-pragmas]
   86 | #pragma unroll
      |         ^
```
You need to guard the pragma with `#ifdef KOKKOS_COMPILER_NVCC`.

Rohit-Kakodkar committed:
Added NVCC compiler guard

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

masterleinad commented:
> You need to guard the pragma with `#ifdef KOKKOS_COMPILER_NVCC`.

We also have `KOKKOS_ENABLE_PRAGMA_UNROLL` that tells you whether `pragma unroll` can be used.

Rombur commented:
I think the PR looks good. 

The question is should this be merged with `chunk_size`. Several backends (`Serial`, `CUDA`, and `HIP`) are currently not using `chunk_size` (at least not in `parallel_for`) so it can be hijacked for the loop unrolling. For `OpenMP`, `OpenMPTarget`, and `OpenACC`, we use `chunk_size` in the loop scheduling. These backends also have `unroll` pragmas. Separating the `chunk_size` and loop unrolling will match nicely for these backends.  `Sycl` is using the `chunk_size` to determine the `workgroup` size. I don't know how loop unrolling would interact with `chunk_size`.

It's clear to me that we should not merge `chunk_size` and loop unrolling. For most of the backends that are currently using `chunk_size`, there is a clear difference between `chunk_size` and loop unrolling and this PR matches what's in the standard of these backends. For the backends that do not use `chunk_size`, it could be a little bit more confusing which parameter to set (but since we are not doing anything now I don't think it matters too much). Moreover to get the expected performance boost, the loop unrolling parameter has to be `constexpr` but `chunk_size` is a runtime parameter. This will make merging these two concepts more complicated.

masterleinad commented:
In addition to the points raised above, `Unroll` seems like a misnomer to me here. What we are really doing is to give every thread more work items. For some reason, this extra loop needs compile-time information to be efficient and we already confirmed that adding `pragm unroll` isn't essential. That's why I'm saying that a name closer to chunk size (maybe even `WorkItemsPerThread`) makes more sense.

Rohit-Kakodkar commented:
I agree with Bruno's point, merging `unroll` and `chunk_size` wouldn't be ideal since I'd expect there to be speed up in OpenMP when using `chunk_size` with `unroll` factors (or atleast the 2 will interact in interesting ways). 

I am not sure if unroll is a misnomer. From my understanding compiler is unrolling the loop since it has access to loop length at compile time even without an explicit `#pragma unroll`. 

For example, looking at [PTX](https://godbolt.org/z/9W878dsz8) for the following code you can see the unrolling. 
```
__global__ 
void loop_unroll_device(int *data, int N, int V) {
    const int tid = threadIdx.x + blockIdx.x * blockDim.x;
    constexpr int loop_unroll_factor = 4;
    if (tid < N) {
        // Unroll the loop
        for (int i = 0; i < loop_unroll_factor; ++i) {
            int index = tid * loop_unroll_factor + i;
            if (index < N) {
                data[index] = V;
            }
        }
    }
}
```

masterleinad commented:
> I am not sure if unroll is a misnomer. From my understanding compiler is unrolling the loop since it has access to loop length at compile time even without an explicit `#pragma unroll`.

Sure the compiler is unrolling the loop but the main point is that this pull request/feature introduces a loop at all.

Rombur commented:
I am open to find a different name but for OpenMP the chunk size is already the number of elements that each thread receives so `WorkItemsPerThread` is not ideal either.

Rohit-Kakodkar commented:
How about something like `StaticBatchSize`, `CompileTimeBatchSize`/`CTBatchSize` ?

Rombur commented:
These names are fine with me. I prefer `StaticBatchSize` but the others work too. What do you think @masterleinad ?

Rohit-Kakodkar committed:
Loop unrool -> StaticBatchSize

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rombur commented:
@masterleinad _ping_

masterleinad commented:
> These names are fine with me. I prefer `StaticBatchSize` but the others work too. What do you think @masterleinad ?

Maybe. I'm just curious what you want to do for, say, `Serial` and `OpenMP`.
For Serial
```
#pragma unroll(batch_size)
  for (typename Policy::member_type i = m_policy.begin(); i < e; ++i) {
      m_functor_reducer.get_functor()(i, update, true);
    }
```
and for `OpenMP`
```
#pragma unroll(batch_size)
    for (auto iwork = ibeg; iwork < iend; ++iwork) {
      exec_work(functor, iwork);
    }
```
? In these two cases, it makes sense to call it something with "unroll" whereas we do more than just unrolling for `Cuda` but change the mapping from hardware threads to work items.

Rohit-Kakodkar commented:
> > These names are fine with me. I prefer `StaticBatchSize` but the others work too. What do you think @masterleinad ?
> 
> Maybe. I'm just curious what you want to do for, say, `Serial` and `OpenMP`. For Serial
> 
> ```
> #pragma unroll(batch_size)
>   for (typename Policy::member_type i = m_policy.begin(); i < e; ++i) {
>       m_functor_reducer.get_functor()(i, update, true);
>     }
> ```
> 
> and for `OpenMP`
> 
> ```
> #pragma unroll(batch_size)
>     for (auto iwork = ibeg; iwork < iend; ++iwork) {
>       exec_work(functor, iwork);
>     }
> ```
> 
> ? In these two cases, it makes sense to call it something with "unroll" whereas we do more than just unrolling for `Cuda` but change the mapping from hardware threads to work items.

My thinking is we would still change the number of  workers that are involved in the `parallel_for`, similar to the way we did in CUDA (Essentially changing the hardware mapping to work_items). With those changes we could implement OpenMP something like this.

```cpp
#pragma omp parallel for
/// when iend -ibeg < batch_size below we'd have single thread working
     for (auto iwork = ibeg; iwork < iend; iwork += batch_size) {
#pragma unroll
          for (auto i = iwork; i < batch_size + iwork; i++)
               exec_worK(functor, i); /// I still need to think if this unroll is cache optimized or do I need a strided unroll? 
     }
```

I think they should be named the same on CPU backends and GPU backends. Since they offer speedup in the same way on both backends - 

1. Assign more work per thread (For Serial this a moot point, but it'd be still more work per "serial" thread, interms of Kokkos symantics)
2. Get better instruction throughput (better ILP in CPUs and more register allocation per thread on GPUs)

masterleinad commented:
> My thinking is we would still change the number of workers that are involved in the `parallel_for`, similar to the way we did in CUDA (Essentially changing the hardware mapping to work_items). With those changes we could implement OpenMP something like this.

In that case, I'm fine with calling it `StaticBatchSize`. We would just need to make sure `chunk_size` isn't defined to a non-default value at the same time.

Rohit-Kakodkar committed:
Fixed tests

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rohit-Kakodkar committed:
Fixed test

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rombur commented on core/src/traits/Kokkos_StaticBatchSizeTrait.hpp:
```diff
@@ -0,0 +1,87 @@
+  struct base_traits {
+    static constexpr bool batch_size_is_defaulted = true;
+
+    using batch_size =
```
```suggestion
    using static_batch_size =
```

Rohit-Kakodkar committed:
fixed kokkos batchsize implementation

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rombur commented on core/src/traits/Kokkos_StaticBatchSizeTrait.hpp:
```diff
@@ -0,0 +1,87 @@
+//
+//@HEADER
+
+#ifndef KOKKOS_KOKKOS_STATIC_BATCH_SIZE_TRAIT_HPP
```
```suggestion
#ifndef KOKKOS_STATIC_BATCH_SIZE_TRAIT_HPP
```

Rombur commented on core/src/traits/Kokkos_StaticBatchSizeTrait.hpp:
```diff
@@ -0,0 +1,87 @@
+//@HEADER
+
+#ifndef KOKKOS_KOKKOS_STATIC_BATCH_SIZE_TRAIT_HPP
+#define KOKKOS_KOKKOS_STATIC_BATCH_SIZE_TRAIT_HPP
```
```suggestion
#define KOKKOS_STATIC_BATCH_SIZE_TRAIT_HPP
```

Rombur commented on core/src/traits/Kokkos_StaticBatchSizeTrait.hpp:
```diff
@@ -0,0 +1,87 @@
+
+}  // end namespace Kokkos::Impl
+
+#endif  // KOKKOS_KOKKOS_S TATIC_BATCH_SIZE_TRAIT_HPP
```
```suggestion
#endif  // KOKKOS_STATIC_BATCH_SIZE_TRAIT_HPP
```

Rombur reviewed (commented):
This looks good. I have a few minor comments.

Rohit-Kakodkar committed:
Update core/src/traits/Kokkos_StaticBatchSizeTrait.hpp

Co-authored-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rohit-Kakodkar committed:
Update core/src/traits/Kokkos_StaticBatchSizeTrait.hpp

Co-authored-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rohit-Kakodkar committed:
Update core/src/traits/Kokkos_StaticBatchSizeTrait.hpp

Co-authored-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rohit-Kakodkar commented:
Let me know if anything needs to change here. Making sure this doesn't get lost.

romintomasetti commented on core/src/traits/Kokkos_StaticBatchSizeTrait.hpp:
```diff
@@ -0,0 +1,87 @@
+  using static_batch_size = StaticBatchSize;
+  using type              = StaticBatchSize<BatchSize>;
+  constexpr static unsigned int batch_size{
+      BatchSize};  // Default batch size is 1
```
Not blocking but it's redundant with the default template argument value.
```suggestion
      BatchSize};
```

romintomasetti commented on core/src/traits/Kokkos_StaticBatchSizeTrait.hpp:
```diff
@@ -0,0 +1,87 @@
+    static constexpr bool batch_size_is_defaulted = true;
+
+    using static_batch_size =
+        Kokkos::Experimental::StaticBatchSize<>;  // Default batch size is 1
```
```suggestion
        Kokkos::Experimental::StaticBatchSize<>;
```

romintomasetti commented on core/unit_test/TestRange.hpp:
```diff
@@ -327,6 +327,71 @@ TEST(TEST_CATEGORY, range_reduce) {
+        Kokkos::create_mirror_view(m_flags);
+
+    // Initialize the flags
+    Kokkos::parallel_for(Kokkos::RangePolicy<ExecSpace>(0, N), *this);
```
This would make your test easier to read:
```suggestion
    Kokkos::deep_copy(m_flags, 0);
```
(and get rid of the untagged call operator below)

romintomasetti commented on core/unit_test/TestRange.hpp:
```diff
@@ -327,6 +327,71 @@ TEST(TEST_CATEGORY, range_reduce) {
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const AtomicAddTag &, const int i) const {
+    // Use atomic add to increment the flag at index i
+    Kokkos::atomic_add(&m_flags(i), 1);
```
In general comments should add more meaningful information :wink:
```suggestion
    Kokkos::atomic_add(&m_flags(i), 1);
```

romintomasetti commented on core/unit_test/TestRange.hpp:
```diff
@@ -327,6 +327,71 @@ TEST(TEST_CATEGORY, range_reduce) {
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const AtomicAddTag &, const int i) const {
```
Tags are assumed to be empty, no need to take them by reference.
```suggestion
  void operator()(const AtomicAddTag, const int i) const {
```

romintomasetti commented on core/unit_test/TestRange.hpp:
```diff
@@ -327,6 +327,71 @@ TEST(TEST_CATEGORY, range_reduce) {
+        Kokkos::RangePolicy<ExecSpace, AtomicAddTag, StaticBatchSize>(0, N),
+        *this);
+
+    Kokkos::parallel_for(
+        Kokkos::RangePolicy<ExecSpace, VerifyAtomicAddTag>(0, N), *this);
+
+    Kokkos::deep_copy(host_flags, m_flags);
+
+    int error_count = 0;
+    for (int i = 0; i < N; ++i) {
+      if (host_flags(i) != 1) {
+        ++error_count;
+      }
+    }
+    ASSERT_EQ(error_count, 0);
```
I'm not sure to understand why you want to check that each flag is set to 1 on both host and device.

It seems the following is sufficient and shorted (not tested but you get the idea)
```suggestion
    bool success = false;
    Kokkos::parallel_reduce(
        Kokkos::RangePolicy<ExecSpace, VerifyTag>(0, N),
        *this
        Kokkos::LAnd<bool>(success)
    );
    ASSERT_TRUE(success);

    ...

KOKKOS_FUNCTION
void operator()(const VerifyTag, bool& current) const {
    current = m_flags(i) == 1 && current;
}
```

romintomasetti commented on core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp:
```diff
@@ -70,22 +71,35 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::Cuda> {
+#if defined(KOKKOS_COMPILER_NVCC)
+#pragma unroll
+#endif
+      for (Member i = 0; ((i < static_cast<Member>(work_stride * batch_size)) &&
+                          (iwork + i < work_end));
+           i += work_stride) {
```
@crtrott @masterleinad @maartenarnst @JBludau

It seems that this can potentially hit unsigned overflow ? (reason why the outer for loop uses a ternary)

JBludau replied:
Yes

romintomasetti replied:
Have you checked that the `nvcc` compiler is able to optimize out your added "inner" loop when the "loop unrolling" policy parameter is set to 1 ?

Rohit-Kakodkar replied:
I think so. The CUDA code collapses back to the one without an unroll factor when unroll factor is 1. https://godbolt.org/z/Wdcdav3En

Rohit-Kakodkar committed:
Removed redundant comments

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rohit-Kakodkar committed:
Fixed tests

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rohit-Kakodkar committed:
Removed not used variable

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

romintomasetti commented on core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp:
```diff
@@ -70,22 +71,35 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::Cuda> {
+             iwork < static_cast<Member>(work_end - work_stride * batch_size)
+                 ? iwork + work_stride * batch_size
+                 : work_end) {
+// Unroll the loop
```
obvious
```suggestion
```

romintomasetti commented on core/unit_test/TestRange.hpp:
```diff
@@ -327,6 +327,61 @@ TEST(TEST_CATEGORY, range_reduce) {
+    Kokkos::parallel_reduce(
+        Kokkos::RangePolicy<ExecSpace>(0, N),
+        KOKKOS_CLASS_LAMBDA(const int i, bool &local_success) {
+          local_success = local_success && (m_flags(i) == 1);
```
Get rid of the `VerifyAtomicAddTag`. No need to launch 2 kernels. In fact I would be in favor of not doing the `printf` at all.
```suggestion
          if (m_flags(i) != 1) {
            Kokkos::printf(
                "TestStaticBatchSize {::test_batch_size_error at %d != %d\n", i,
                m_flags(i));
            }
          local_success = local_success && (m_flags(i) == 1);
```

romintomasetti commented on core/unit_test/TestRange.hpp:
```diff
@@ -327,6 +327,61 @@ TEST(TEST_CATEGORY, range_reduce) {
+  struct AtomicAddTag {};
+  struct VerifyAtomicAddTag {};
+
+  int N;
+
+  TestStaticBatchSize(const size_t N_)
```
Make sure your types agree.
```suggestion
  size_t N;

  TestStaticBatchSize(const size_t N_)
```

Rohit-Kakodkar committed:
Minor fixes

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

romintomasetti commented on core/unit_test/TestRange.hpp:
```diff
@@ -327,6 +327,63 @@ TEST(TEST_CATEGORY, range_reduce) {
+  void operator()(const AtomicAddTag, const int i) const {
+    Kokkos::atomic_add(&m_flags(i), 1);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const VerifyAtomicAddTag, const int i) const {
+    if (m_flags(i) != 1) {
+      Kokkos::printf(
+          "TestStaticBatchSize {::test_batch_size_error at %d != %d\n", i,
+          m_flags(i));
+    }
+  }
```
Dead code.

romintomasetti commented on core/unit_test/TestRange.hpp:
```diff
@@ -327,6 +327,63 @@ TEST(TEST_CATEGORY, range_reduce) {
+  view_type result_view;
+
+  struct AtomicAddTag {};
+  struct VerifyAtomicAddTag {};
```
Dead code.

romintomasetti commented on core/unit_test/TestRange.hpp:
```diff
@@ -327,6 +327,63 @@ TEST(TEST_CATEGORY, range_reduce) {
+    ASSERT_TRUE(success);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const AtomicAddTag, const int i) const {
+    Kokkos::atomic_add(&m_flags(i), 1);
+  }
```
Once you get rid of `VerifyAtomicAddTag`, you can get rid of the tag dispatch.
```suggestion
  KOKKOS_INLINE_FUNCTION
  void operator()(const int i) const {
    Kokkos::atomic_add(&m_flags(i), 1);
  }
```

Rohit-Kakodkar replied:
I moved the lambda functor into the tagged operator to maintain consistency with other tests in this suite. Please let me know if you'd prefer me not to use tagged dispatch.

romintomasetti replied:
I cannot strongly oppose to keeping consistency 😉

romintomasetti commented on core/unit_test/TestRange.hpp:
```diff
@@ -382,4 +439,47 @@ TEST(TEST_CATEGORY, large_parallel_for_reduce) {
+  ASSERT_TRUE(Kokkos::Experimental::StaticBatchSize<4>::batch_size == 4);
+}
+
+TEST(TEST_CATEGORY, range_static_batch_size) {
+  {
+    TestStaticBatchSize<TEST_EXECSPACE,
+                        Kokkos::Experimental::StaticBatchSize<1>>
+        f(1024);
+    f.test_batch_size();
+  }
```
This is a suggestion but I'm fine not doing it.

I would propose you pass an execution space instance to avoid all the additional fences `Kokkos` is adding otherwise.

```suggestion
TEST(TEST_CATEGORY, range_static_batch_size) {
  const TEST_EXECSPACE exec{};
  {
    TestStaticBatchSize<TEST_EXECSPACE,
                        Kokkos::Experimental::StaticBatchSize<1>>
        f(exec, 1024);
    f.test_batch_size(exec);
  }
```

Rohit-Kakodkar committed:
Moved lambda to tagged operator

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rohit-Kakodkar committed:
Moved lambda to tagged operator

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

Rohit-Kakodkar commented:
@romintomasetti, feel free to share any other suggestions you might have!

crtrott commented:
I think I am ok with it. Though I am wondering if a strategy with an if constexpr for the unrolling might be better. I.e. sidestep the second loop completely if there is no unrolling requested. Anyway I started testing again.,

Rombur commented:
>  Anyway I started testing again.,

Any update on this @crtrott

Rombur commented:
@crtrott _ping_

masterleinad reviewed (commented):
I'm OK with only implementing the feature for Cuda right (since it's in `Experimental`) now but would be prefer an implementation for all bcakends at once. I would approve after resolving the discussion in https://github.com/kokkos/kokkos/pull/8164/files#r2263249878 regardimg potetial overflow (I would also just use a ternary).

Rohit-Kakodkar commented:
> I'm OK with only implementing the feature for Cuda right (since it's in `Experimental`) now but would be prefer an implementation for all bcakends at once. I would approve after resolving the discussion in https://github.com/kokkos/kokkos/pull/8164/files#r2263249878 regardimg potetial overflow (I would also just use a ternary).

I was planning to submit separate PRs for different backends to keep each PR small. Regarding your second comment, what would be the workaround for the overflow error? @romintomasetti @masterleinad

masterleinad commented:
> Regarding your second comment, what would be the workaround for the overflow error?

Something like
```C++
for (Member i = 0; ((i < static_cast<Member>(work_stride * batch_size)) &&
                          (i < work_end - iwork));
           i = (i < (work_end - work_stride - iwork)) ? i + work_stride) : work_end - iwork) {
```
should do the trick.

Rohit-Kakodkar committed:
Fixed overflow error

Signed-off-by: Rohit Kakodkar <kakodkar.rohit@gmail.com>

crtrott commented:
I did a little bit of perf testing too and yeah so far looks fine. We will see if someone downstream reports issues.

science-enthusiast commented:
> Something like
> 
> ```c++
> for (Member i = 0; ((i < static_cast<Member>(work_stride * batch_size)) &&
>                           (i < work_end - iwork));
>            i = (i < (work_end - work_stride - iwork)) ? i + work_stride) : work_end - iwork) {
> ```
> 
> should do the trick.

This is causing CI failures in Jenkins: https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7675/74/pipeline-overview/?selected-node=330

dalg24 commented:
> > Something like
> > ```c++
> > for (Member i = 0; ((i < static_cast<Member>(work_stride * batch_size)) &&
> >                           (i < work_end - iwork));
> >            i = (i < (work_end - work_stride - iwork)) ? i + work_stride) : work_end - iwork) {
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > should do the trick.
> 
> This is causing CI failures in Jenkins: https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7675/74/pipeline-overview/?selected-node=330

```
/var/jenkins/workspace/Kokkos_PR-7675/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp:77:19: error: comparison of integers of different signs: 'Member' (aka 'int') and 'unsigned int' [clang-diagnostic-sign-compare]
   77 |            i = (i < (work_end - work_stride - iwork)) ? i + work_stride
      |                 ~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Rohit-Kakodkar commented:
> > > Something like
> > > ```c++
> > > for (Member i = 0; ((i < static_cast<Member>(work_stride * batch_size)) &&
> > >                           (i < work_end - iwork));
> > >            i = (i < (work_end - work_stride - iwork)) ? i + work_stride) : work_end - iwork) {
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > should do the trick.
> > 
> > 
> > This is causing CI failures in Jenkins: https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7675/74/pipeline-overview/?selected-node=330
> 
> ```
> /var/jenkins/workspace/Kokkos_PR-7675/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp:77:19: error: comparison of integers of different signs: 'Member' (aka 'int') and 'unsigned int' [clang-diagnostic-sign-compare]
>    77 |            i = (i < (work_end - work_stride - iwork)) ? i + work_stride
>       |                 ~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```

Thanks, pushing a fix rn.

science-enthusiast commented on core/unit_test/TestRange.hpp:
```diff
@@ -327,6 +327,55 @@ TEST(TEST_CATEGORY, range_reduce) {
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const AtomicAddTag, const int i) const {
+    Kokkos::atomic_add(&m_flags(i), 1);
```
Hi @Rohit-Kakodkar
Could you clarify, why an atomic_add is needed here? Why not a simple add or assignment with respect to the element of the View? I need to write similar tests for MDRangePolicy in [8629](https://github.com/kokkos/kokkos/pull/8629).

Rohit-Kakodkar replied:
It's been a while. 

AtomicAdd to would test against the rare case where multiple threads would write to same location at the same time. Regular update or assignment might miss that race condition. i.e. atomicadd would result in 2, assignment would definitely result 1 (regardless of how many threads accessed that value) & simple add could result in 1 (if a thread picks up the value while another thread is in process of updating it)

science-enthusiast replied:
Thank you for your quick response.

Yes, the atomic_add will be a good way to catch incorrect concurrent execution in a future update of the code base. It will be good to keep this in mind, in general, while writing tests.
